velox/exec/HashTable.cpp

```diff
@@ -1750,18 +1750,62 @@ void HashTable<ignoreNullKeys>::prepareJoinTable(
   }
 }
 
+template <bool ignoreNullKeys>
+inline uint64_t HashTable<ignoreNullKeys>::joinProjectedVarColumnsSize(
+    const std::vector<vector_size_t>& columns,
+    const char* row) const {
+  uint64_t totalBytes{0};
+  for (const auto& column : columns) {
+    if (!rows_->columnTypes()[column]->isFixedWidth()) {
+      totalBytes += rows_->variableSizeAt(row, column);
+    }
+  }
+  return totalBytes;
+}
+
+template <bool ignoreNullKeys>
+inline uint64_t HashTable<ignoreNullKeys>::joinProjectedFixColumnsSize(
+    const std::vector<vector_size_t>& columns) const {
+  uint64_t totalBytes{0};
+  for (const auto& column : columns) {
+    const auto& type = rows_->columnTypes()[column];
+    VELOX_CHECK(type->isFixedWidth());
+    totalBytes += rows_->fixedSizeAt(column);
+  }
+  return totalBytes;
+}
+
 template <bool ignoreNullKeys>
 int32_t HashTable<ignoreNullKeys>::listJoinResults(
     JoinResultIterator& iter,
     bool includeMisses,
     folly::Range<vector_size_t*> inputRows,
-    folly::Range<char**> hits) {
+    folly::Range<char**> hits,
+    uint64_t maxBytes) {
   VELOX_CHECK_LE(inputRows.size(), hits.size());
-  if (!hasDuplicates_) {
-    return listJoinResultsNoDuplicates(iter, includeMisses, inputRows, hits);
+  std::vector<vector_size_t> varSizeColumns;
+  std::vector<vector_size_t> fixedSizeColumns;
+  for (const auto column : iter.listColumns) {
```

Comment:
I think we could initialize this once and keep two lists in join result iterator? one is fixedSizeListJoinColumns, and varSizeListJoinColumns? And for fixed size list columns, we could simply store the sum which is the same for all the rows? thanks!